### ISSUE TITLE
Increment arrows fail to satisfy required constraint on Firefox

### ISSUE THREAD

**Issue by TeraMike** [UserOfficeProject/stfc-user-office-project#281]
Please refer to [this Teams thread](https://teams.microsoft.com/l/message/19:f0afa005ea5b474e83109a55c93c16f2@thread.skype/1636985560790?tenantId=3f66361c-a87e-4158-8f61-99e82db3cac8&groupId=403a668a-985f-445b-a4e3-b3a3070bb25b&parentMessageId=1636985560790&teamName=ISIS%20Computing%20Division%20Staff&channelName=Proposals%20and%20Outcomes&createdTime=1636985560790) for the discussion/identification of the issue.  
Identified as a part of #235.

### Description  
When using Firefox, if you enter a value into a required* increment field (i.e.: you can use arrows to increment a value instead of typing) using only the increment arrows, then you cannot save and continue until you have selected and de-selected the field. In other words, the increment buttons alone do not appear to properly save a value to indicate that a required field has been filled out.

This issue does not occur on Chrome.  
More testing is likely required (does this issue occur on other browsers, or other increment widgets, etc.) for a complete picture.

### To reproduce:  
- Use Firefox 
- Find a required field with an increment (e.g.: (2) Experiment on the calls on prod as in screenshots)
- Use the arrows to pick any value and DO NOT click to type a value
- Click `Save and Continue`

---

### Screenshots

Both boxes failing on Firefox without being selected:
![image](https://user-images.githubusercontent.com/57662035/141811730-6d4b8c77-8bad-4e2a-b936-c16621ee1a5a.png)

One box (years) selected and other (weeks) with only increment arrows used:
![image](https://user-images.githubusercontent.com/57662035/141806854-f5bb3ad9-def5-4f93-829a-576c7d8ace48.png)  
See that only the latter box fails.

These are the arrows I am referring to: 
![image](https://user-images.githubusercontent.com/57662035/142865657-b42509b8-63d3-43a2-9792-56f6310df6d3.png)

**Comment by TickleThePanda** [comment]
I think this is something that Emma Gozzard saw in one of the bugs before, but we were unable to replicate it.

**Comment by WhittyPM** [comment]
@TickleThePanda @TeraMike As Panda said this was the same issue Emma experienced. Some of the issues she had were related to caching...

**Comment by TeraMike** [comment]
I'm able to replicate this consistently on Firefox. I cleared the cache and this didn't seem to change the behaviour.  
It also happens if you have a saved page (e.g.: with 'Octopus' and 'No' for resubmissions) that you return to: incrementing the values won't un-grey the Save button, indicating that no changes have been made.

The arrows function fine on Chrome and Edge so it must be related to Firefox somehow, I figure. But if others cannot replicate it then I'm not sure what it would be. It seemed reproducible in the original thread yesterday at least.

**Comment by TickleThePanda** [comment]
I think the caching issue was separate to this, here.

**Comment by keiranjprice101** [comment]
@TeraMike 
I could not produce this locally and I can no longer produce this on dev either. Can you confirm that you can still produce this before I close it
![arrows](https://user-images.githubusercontent.com/44777678/142599199-e2934c28-d66e-4e5d-bcd1-a8d6d8d913c9.gif)
?

**Comment by TeraMike** [comment]
The issue isn't with typing the values in, it's using the arrows.

If you enter a value with the arrows _without_ typing it you cannot proceed. As in, you use the arrows to enter a value without selecting the box -- it's not the whole field that's the issue, but the arrows specifically, I believe.

I'm not able to record my screen to get a gif for you unfortunately but I can reproduce this 100% of the time on dev and prod. I can show you on a call if that would be useful.

**Comment by keiranjprice101** [comment]
@TeraMike Yes arrows are used in the recordings, not typing. Have you retested this since the latest release?

**Comment by TeraMike** [comment]
> @TeraMike Yes arrows are used in the recordings, not typing. Have you retested this since the latest release?

In the gif you posted before you type the numbers in instead of incrementing them with the arrows.  
![image](https://user-images.githubusercontent.com/57662035/142857682-764a278f-cea9-4531-8caf-018362735650.png)

The issue is particularly with the increment arrows (I think the other term is spinner), not using the keyboard to navigate the page. If you click the arrows/spinners without selecting the box (highlighted below), the values are not registered properly. It's a bit narrow admittedly but very reproducible. Selecting the box in some capacity (clicking or tabbing) will resolve the issue and allow you to progress.

![image](https://user-images.githubusercontent.com/57662035/142859479-1941dda4-4fcc-43d5-8e8c-3d4db78c9f2b.png)

**Comment by keiranjprice101** [comment]
@TeraMike I think I see where the confusion lies. In the Gif I used the arrow keys on the keyboard, but it refers to the arrows within the spinner 🤦 My mistake

**Comment by TeraMike** [comment]
> @TeraMike I think I see where the confusion lies. In the Gif I used the arrow keys on the keyboard, but it refers to the arrows within the spinner 🤦 My mistake

I probably could have explained myself better, no worries :)

**Comment by ACLay** [comment]
I've tried following the invalid event through firefox's dev tools. Based on the presence of a string that reads 'Please specify unit', which seems to be unique in the UOP codebase, I'm pretty sure the failing validation is in https://github.com/UserOfficeProject/user-office-lib/blob/f474195182dc68451dc75d69d57a329c459d2695/packages/validation/src/Questionary/numberInput.ts

**Comment by ACLay** [comment]
Using the react dev tools, in firefox it seems that the arrows are updating the hooks state of the component, but this isn't propagating up to the component's value property, while in chrome the value does get updated
![image](https://user-images.githubusercontent.com/2938284/148812839-7fad1db8-4333-4517-aa2d-0c72ef10b66b.png)

**Comment by ACLay** [comment]
In `QuestionaryComponentNumberInput.tsx` we use `onChange` and `onBlur` (loss of focus) events to manage the component's state when the textbox is interacted with. I've confirmed by adding log statements in those event listeners that firefox isn't triggering the `onBlur` event, while chrome is. It turns out that this is because [firefox does not assign focus to the textbox when using the arrows](https://bugzilla.mozilla.org/show_bug.cgi?id=1012818) so there's no focus to lose and the listener never triggers. That's been open with little attention since 2014, so we'll need a small js fix as an upstream fix seems unlikely.
